gdb/gdb.go

```diff
@@ -1,198 +1,285 @@
 // Package gdb interacts with thegamedb.net's API.
-//
-// Example:
-//  resp, err := gdb.GetGame(GGReq{ID: "5"})
 package gdb
 
 import (
 	"context"
-	"encoding/xml"
 	"fmt"
 	"net/http"
-	"net/url"
-)
+	"strconv"
+	"sync"
+
+	"github.com/antihax/optional"
 
-const (
-	gdbURL  = "http://legacy.thegamesdb.net"
-	ggPath  = "/api/GetGame.php"
-	gglPath = "/api/GetGamesList.php"
+	gamesdb "github.com/J-Swift/thegamesdb-swagger-client-go"
 )
 
-// GGLReq represents a request for a GetGameList command.
-type GGLReq struct {
-	Name     string
-	Platform string
-	Genre    string
+var apiClient = gamesdb.NewAPIClient(gamesdb.NewConfiguration())
+
+// Publishers
+
+type publishers struct {
+	mux        sync.Mutex
+	publishers *map[string]gamesdb.Publisher
 }
 
-// GGLResp represents the response of a GetGameList command.
-type GGLResp struct {
-	XMLName xml.Name
-	Game    []GameTrunc
-	Err     string `xml:",chardata"`
+var publishersCache = publishers{}
+
+func getPublishers(ctx context.Context, apikey string) map[string]gamesdb.Publisher {
+	pubs, resp, err := apiClient.PublishersApi.Publishers(ctx, apikey)
+
+	if err != nil || resp.StatusCode != 200 {
+		return make(map[string]gamesdb.Publisher)
+	}
+
+	return pubs.Data.Publishers
 }
 
-// GameTrunc is used to parse the GetGamesList's <Game> tag.
-type GameTrunc struct {
-	ID          string `xml:"id"`
-	GameTitle   string
-	ReleaseDate string
-	Platform    string
+func getCachedPublishers(ctx context.Context, apikey string) map[string]gamesdb.Publisher {
```

Comment:
These could be something like the following to make them less verbose when used below and remove the odd looking pointer to a map. But they are probably okay as is.

```
func (p *pubplisher) get(ctx context.Context, apikey string) map[string]gamesdb.Publisher {
  ...
}
```

or

```
func (p *pubplisher) get(ctx context.Context, apikey string, id string) gamesdb.Publisher {
  ...
}
```

or

```
func (p *pubplisher) get(ctx context.Context, apikey string, id string) ParsedPublisher {
  ...
}
```

Reply:
This caching was really quickly hacked together toward the end, so definitely could use some cleanup.

I think I did the pointer to a distinguish cache-miss vs cache-hit-but-empty-value. I also generally avoid extensions but I don't know how golang idioms feel about it.